File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> result;
        if(p.length() > s.length())
        {
            return result;
        }
        vector<int> pmp(26,0);
        vector<int> smp(26,0);
        for(int i=0;i<p.length();i++)
        {
            char ch = p[i];
            pmp[ch - 'a']++;
        }
        int n = p.length()-1;
        char c;
        for(int j=0;j<n;j++)
        {
            c = s[j];
            smp[c - 'a']++;   
        }

        int k=0;
        for(int i=n;i<s.size();i++)
        {
            c = s[i];
            smp[c - 'a']++;
            if(smp == pmp)
            {
                result.push_back(k);
            }
            smp[s[k]-'a']--;
            k++;
        }

        return result;
    }
};

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.py
class Solution(object):
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        if(len(p)>len(s)):
            return []
        l1 = [0]*26
        for ch in p:
            i = ord(ch)-ord('a')
            l1[i] = l1[i]+1
        
        l2 = [0]*26
        i=0
        j=0
        ans = []
        while(j<len(p)-1):
            idx = ord(s[j])-ord('a')
            l2[idx] +=1
            j +=1
        
        while(j<len(s)):
            idx = ord(s[j])-ord('a')
            l2[idx] +=1
            if(l1 == l2):
                ans.append(i)
            idx1 = ord(s[i])-ord('a')
            l2[idx1] -=1
            i +=1
            j +=1
        return ans

        